mlir/lib/ExecutionEngine/ROCm/BackendUtils.cpp

```diff
@@ -41,17 +41,94 @@
 
 #include <mutex>
 #include <numeric>
+#include <string.h>
 
 using namespace mlir;
 
 static constexpr const char kRunnerProgram[] = "mlir-rocm-runner";
 static constexpr const char kRocmAgentEnumerator[] = "rocm_agent_enumerator";
+static constexpr const char gcnArchDelimiter[] = ":";
+static constexpr const char targetTriple[] = "amdgcn-amd-amdhsa";
+
+namespace {
+// TODO: remove this once the rocm_agent_enumerator is ready
+void getGpuGCNArchName(hipDevice_t device, std::string &gcnArchName) {
+  hipDeviceProp_t props;
+  hipError_t result = hipGetDeviceProperties(&props, device);
+  if (result != hipSuccess) {
+    gcnArchName = "";
+    return;
+  }
+
+  const char *pArchName = props.gcnArchName;
+  gcnArchName.assign(pArchName);
+}
+
+// This function convert the arch name to target features:
+// gfx908:sramecc+:xnack- to +sramecc,-xnack
+std::string getFeaturesFromArchName(const std::string &gcnArchName) {
+  std::string features;
+
+  // First step: get rid of the arch name in feature string
+  std::size_t firstSeperatorLoc = gcnArchName.find(gcnArchDelimiter);
+  if (firstSeperatorLoc == std::string::npos) {
+    return features;
+  }
+  std::string gcnArchFeature = gcnArchName.substr(firstSeperatorLoc + 1);
+
+  // Second step: put each feature name to the vector
+  std::string token;
+  std::vector<std::string> featureTokens;
+  auto convertFeatureToken = [](std::string &token) {
+    if (token.back() != '+' && token.back() != '-') {
+      llvm::errs() << "malformed token: " << token << ", must end with +/-.\n";
+      return token;
+    }
+    token.insert(token.begin(), token.back());
+    token.pop_back();
+    return token;
+  };
+
+  size_t featureEnd = 0;
+  while ((featureEnd = gcnArchFeature.find(gcnArchDelimiter)) !=
+         std::string::npos) {
+    token = gcnArchFeature.substr(0, featureEnd);
+    featureTokens.push_back(convertFeatureToken(token));
+    gcnArchFeature.erase(0, featureEnd + strlen(gcnArchDelimiter));
+  }
+  featureTokens.push_back(convertFeatureToken(gcnArchFeature));
+
+  // Third step: join processed token back to feature string
+  std::string gcnFeatureDelimiter = ",";
+  features = std::accumulate(
+      std::next(featureTokens.begin()), featureTokens.end(), featureTokens[0],
```

Comment:
This'll reference unset memory when there are no features. Please add a check.

Reply:
How so? std::next of the empty vector will be equivalent to end iterator, making accumulate to not preceed.

Reply:
`featureTokens[0]` on the empty vector will read off the end

Reply:
Fair enough. The empty vector load will only happen when a `gfx908:` is a input. Though I don't know how likely that could be, still good to double check to prevent a malformed string crash us.

Reply:
I agree that the case isn't likely, but people write bugs in their code (or, in some cases, are actively evil) and it's worth ensuring those are guarded against.